[View in Colaboratory](https://colab.research.google.com/github/afarane/Analyze_NYC_FlighData/blob/master/DS_AF_Upx_NYC_FlightData.ipynb)

## 1. Load Dataset and Overview

In [0]:
import pandas as pd
import numpy as np

In [5]:
# 1.1: Load Dataset : ----- 
#train_data = pd.read_csv('flight_data.csv', error_bad_lines=False)
flighData = pd.read_csv('flight_data.csv')
flighData.head()


year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0  2013      1    1     517.0             515        2.0     830.0   
1  2013      1    1     533.0             529        4.0     850.0   
2  2013      1    1     542.0             540        2.0     923.0   
3  2013      1    1     544.0             545       -1.0    1004.0   
4  2013      1    1     554.0             600       -6.0     812.0   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             819       11.0      UA  1545.0  N14228    EWR  IAH     227.0   
1             830       20.0      UA  1714.0  N24211    LGA  IAH     227.0   
2             850       33.0      AA  1141.0  N619AA    JFK  MIA     160.0   
3            1022      -18.0      B6   725.0  N804JB    JFK  BQN     183.0   
4             837      -25.0      DL   461.0  N668DN    LGA  ATL     116.0   

   distance  hour  minute         time_hour  
0    1400.0   5.0    15.0  01-01-2013 05:00  
1    1416.0   5.0    29.0  01-01-2013 05:00  
2    1089.0   5.0    40.0  01-01-2013 05:00  
3    1576.0   5.0    45.0  01-01-2013 05:00  
4     762.0   6.0     0.0  01-01-2013 06:00

In [3]:
# 1.2: ProfileReport of Dataset : -----
#!pip install pandas-profiling
#import pandas_profiling 
#pfr = pandas_profiling.ProfileReport(flighData)
#pfr.to_file("DS_AF_Upx_NYC_FlightData.html")
#pfr

ModuleNotFoundError: ignored

In [0]:
# 1.3: shape : -----
flighData.shape

(336776, 19)

In [0]:
# 1.5:. Numerical Columns : -----
flighData.describe(include=['number'])

year          month            day       dep_time  sched_dep_time  \
count  336776.0  336776.000000  336776.000000  328521.000000   336776.000000   
mean     2013.0       6.548510      15.710787    1349.109947     1344.254840   
std         0.0       3.414457       8.768607     488.281791      467.335756   
min      2013.0       1.000000       1.000000       1.000000      106.000000   
25%      2013.0       4.000000       8.000000     907.000000      906.000000   
50%      2013.0       7.000000      16.000000    1401.000000     1359.000000   
75%      2013.0      10.000000      23.000000    1744.000000     1729.000000   
max      2013.0      12.000000      31.000000    2400.000000     2359.000000   

           dep_delay       arr_time  sched_arr_time      arr_delay  \
count  328521.000000  328063.000000   336776.000000  327346.000000   
mean       12.639070    1502.054999     1536.380220       6.895377   
std        40.210061     533.264132      497.457142      44.633292   
min       -43.000000       1.000000        1.000000     -86.000000   
25%        -5.000000    1104.000000     1124.000000     -17.000000   
50%        -2.000000    1535.000000     1556.000000      -5.000000   
75%        11.000000    1940.000000     1945.000000      14.000000   
max      1301.000000    2400.000000     2359.000000    1272.000000   

              flight       air_time       distance           hour  \
count  336776.000000  327346.000000  336776.000000  336776.000000   
mean     1971.923620     150.686460    1039.912604      13.180247   
std      1632.471938      93.688305     733.233033       4.661316   
min         1.000000      20.000000      17.000000       1.000000   
25%       553.000000      82.000000     502.000000       9.000000   
50%      1496.000000     129.000000     872.000000      13.000000   
75%      3465.000000     192.000000    1389.000000      17.000000   
max      8500.000000     695.000000    4983.000000      23.000000   

              minute  
count  336776.000000  
mean       26.230100  
std        19.300846  
min         0.000000  
25%         8.000000  
50%        29.000000  
75%        44.000000  
max        59.000000

In [0]:
# 1.6: Categorical Columns : -----
flighData.describe(include=['object'])

carrier tailnum  origin    dest         time_hour
count   336776  334264  336776  336776            336776
unique      16    4043       3     105              6936
top         UA  N725MQ     EWR     ORD  13-09-2013 08:00
freq     58665     575  120835   17283                94

## 2. Check Unique Values :

In [4]:
flighData.flight.nunique() == flighData.shape[0]
flighData.tailnum.nunique() == flighData.shape[0]

NameError: ignored

## 3. Check & Fix Missing data

In [0]:
# 3.1: Overview : -----
flighData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 336776 entries, 0 to 336775
Data columns (total 19 columns):
year              336776 non-null int64
month             336776 non-null int64
day               336776 non-null int64
dep_time          328521 non-null float64
sched_dep_time    336776 non-null int64
dep_delay         328521 non-null float64
arr_time          328063 non-null float64
sched_arr_time    336776 non-null int64
arr_delay         327346 non-null float64
carrier           336776 non-null object
flight            336776 non-null int64
tailnum           334264 non-null object
origin            336776 non-null object
dest              336776 non-null object
air_time          327346 non-null float64
distance          336776 non-null int64
hour              336776 non-null int64
minute            336776 non-null int64
time_hour         336776 non-null object
dtypes: float64(5), int64(9), object(5)
memory usage: 48.8+ MB


In [0]:
# 3.2. Check Null : -----
flighData.isnull().sum()


year                 0
month                0
day                  0
dep_time          8255
sched_dep_time       0
dep_delay         8255
arr_time          8713
sched_arr_time       0
arr_delay         9430
carrier              0
flight               0
tailnum           2512
origin               0
dest                 0
air_time          9430
distance             0
hour                 0
minute               0
time_hour            0
dtype: int64

In [0]:
len(flighData[flighData['arr_time'].isnull() & flighData['dep_time'].isnull()])

8255

Note 1:
- There are 8255 rows in which both arrival & departure time is not available. We can drop above rows from dataset!

In [0]:
flighData.dropna(subset=['dep_time', 'arr_time'], how='all',inplace=True)
flighData.isnull().sum()

year                 0
month                0
day                  0
dep_time             0
sched_dep_time       0
dep_delay            0
arr_time           458
sched_arr_time       0
arr_delay         1175
carrier              0
flight               0
tailnum              0
origin               0
dest                 0
air_time          1175
distance             0
hour                 0
minute               0
time_hour            0
dtype: int64

Note 2:
- There are 458 Flights for which we have arr_time is missing but we have dep_time. That means these flights actually started their journey . Now we need to find their arr_time and fill it accuratly in 'arr_time'
- We can fill these rows by adding tavel time in dep_time . sched_arr_time - sched_dep_time

In [0]:
# as per HH:MM:SS Format

flighData["sched_dep_time"]= pd.to_numeric(flighData['sched_dep_time']).fillna(0).astype(np.int64).apply(lambda x: '{0:0>4}'.format(x))
flighData["sched_dep_time"]= pd.to_timedelta(flighData["sched_dep_time"].str[:2]+":"+ flighData["sched_dep_time"].str[-2:]+":00")

flighData["sched_arr_time"]= pd.to_numeric(flighData['sched_arr_time']).fillna(0).astype(np.int64).apply(lambda x: '{0:0>4}'.format(x))
flighData["sched_arr_time"]= pd.to_timedelta(flighData["sched_arr_time"].str[:2]+":"+ flighData["sched_arr_time"].str[-2:]+":00")

temp_arr_time = pd.concat([flighData["sched_dep_time"],flighData["sched_arr_time"]], axis=1, keys=['sched_dep_time','sched_arr_time'])
temp_arr_time.head()

sched_dep_time sched_arr_time
0       05:15:00       08:19:00
1       05:29:00       08:30:00
2       05:40:00       08:50:00
3       05:45:00       10:22:00
4       06:00:00       08:37:00

In [0]:
flighData["dep_time_new"]= pd.to_numeric(flighData['dep_time']).fillna(0).astype(np.int64).apply(lambda x: '{0:0>4}'.format(x))
flighData["dep_time_new"]= pd.to_timedelta(flighData["dep_time_new"].str[:2]+":"+ flighData["dep_time_new"].str[-2:]+":00")

flighData["arr_time_new"]= pd.to_numeric(flighData['arr_time']).fillna(0).astype(np.int64).apply(lambda x: '{0:0>4}'.format(x))
flighData["arr_time_new"]= pd.to_timedelta(flighData["arr_time_new"].str[:2]+":"+ flighData["arr_time_new"].str[-2:]+":00")

temp_arr_time = pd.concat([flighData["dep_time"],flighData["dep_time_new"],flighData["arr_time"],flighData["arr_time_new"]], axis=1, keys=['dep_time','dep_time_new','arr_time','arr_time_new'])
temp_arr_time.head()

dep_time dep_time_new  arr_time arr_time_new
0     517.0     05:17:00     830.0     08:30:00
1     533.0     05:33:00     850.0     08:50:00
2     542.0     05:42:00     923.0     09:23:00
3     544.0     05:44:00    1004.0     10:04:00
4     554.0     05:54:00     812.0     08:12:00

In [0]:
flighData['time_hour']=pd.to_datetime(flighData['time_hour'])
flighData.head()

year  month  day  dep_time sched_dep_time  dep_delay  arr_time  \
0  2013      1    1     517.0       05:15:00        2.0     830.0   
1  2013      1    1     533.0       05:29:00        4.0     850.0   
2  2013      1    1     542.0       05:40:00        2.0     923.0   
3  2013      1    1     544.0       05:45:00       -1.0    1004.0   
4  2013      1    1     554.0       06:00:00       -6.0     812.0   

  sched_arr_time  arr_delay carrier    ...      origin dest air_time distance  \
0       08:19:00       11.0      UA    ...         EWR  IAH    227.0     1400   
1       08:30:00       20.0      UA    ...         LGA  IAH    227.0     1416   
2       08:50:00       33.0      AA    ...         JFK  MIA    160.0     1089   
3       10:22:00      -18.0      B6    ...         JFK  BQN    183.0     1576   
4       08:37:00      -25.0      DL    ...         LGA  ATL    116.0      762   

   hour  minute           time_hour  dep_time_new arr_time_new sched_diff  
0     5      15 2013-01-01 05:00:00      05:17:00     08:30:00   03:04:00  
1     5      29 2013-01-01 05:00:00      05:33:00     08:50:00   03:01:00  
2     5      40 2013-01-01 05:00:00      05:42:00     09:23:00   03:10:00  
3     5      45 2013-01-01 05:00:00      05:44:00     10:04:00   04:37:00  
4     6       0 2013-01-01 06:00:00      05:54:00     08:12:00   02:37:00  

[5 rows x 22 columns]

In [0]:
missing_arr_tme_index = flighData[flighData['arr_time'].isnull() == True]